nixos filesystems: unify special filesystems handling

abbradar opened the pull request:
###### Motivation for this change

This makes various mount points have the same options as [systemd wants](https://github.com/NixOS/systemd/blob/nixos-v230/src/core/mount-setup.c#L72). This also seems to be more secure to me.
###### Things done
- [x] Tested using sandboxing
  ([nix.useChroot](http://nixos.org/nixos/manual/options.html#opt-nix.useChroot) on NixOS,
    or option `build-use-chroot` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

I don't like the repetiveness very much -- any ideas how to improve this?

cc @dezgeg

abbradar commented:
I've redone this to unify options of all special filesystems in all places. Quoting commit message:

Add new internal config option `boot.specialFileSystems` is added which holds a
shell snippet with filesystems that need to be mounted very early (i.e. in
initrd).  Filesystems are listed via calls to an undefined `specialMount`
function, which is expected to be caller-specific. This option is used by
stage-1, stage-2 and activation script to set up and remount those filesystems.
Options for them are updated according to systemd defaults.

edolstra commented on nixos/modules/tasks/filesystems.nix:
```diff
@@ -155,6 +174,17 @@ in
 
+    boot.specialFileSystems = mkOption {
+      type = types.listOf (types.submodule coreFileSystemOpts);
+      apply = makeSpecialMounts;
```
Don't use `apply`, it's ugly. If you really must generate a shell script (rather than say a file in /etc/fstab format), it's better to have a value like `sytem.build.specialFilesystemMountScript` derived from `boot.specialFileSystems`.

dezgeg replied:
I thought about the fstab format as well, but that probably won't work too well since someone has to create the mountpoint directories beforehand (and busybox doesn't seem to support `x-mount.mkdir[=mode]` like util-linux).

edolstra commented:
Hm, why not use the regular `fileSystems` option?

dezgeg commented on nixos/modules/system/activation/activation-script.nix:
```diff
@@ -154,9 +154,15 @@ in
 
     system.activationScripts.tmpfs =
       ''
-        ${pkgs.utillinux}/bin/mount -o "remount,size=${config.boot.devSize}" none /dev
```
I don't think this is needed, given that during remount only the `size=` should be changing.

abbradar replied:
As I've discovered, it also drops all earlier options if not given them again. 
-- 
Nikolay. 

Sent from a mobile device, please excuse my brevity.

dezgeg replied:
I can't reproduce that:

````
[tmtynkky@kbuilder:~]$ mkdir /tmp/test

[tmtynkky@kbuilder:~]$ sudo mount -t tmpfs tmpfs /tmp/test

[tmtynkky@kbuilder:~]$ mount | grep tmp.test
tmpfs on /tmp/test type tmpfs (rw,relatime)

[tmtynkky@kbuilder:~]$ sudo mount -o remount,size=4G /tmp/test

[tmtynkky@kbuilder:~]$ mount | grep tmp.test
tmpfs on /tmp/test type tmpfs (rw,relatime,size=4194304k)

[tmtynkky@kbuilder:~]$ sudo mount -o remount,nr_inodes=100k /tmp/test

[tmtynkky@kbuilder:~]$ mount | grep tmp.test
tmpfs on /tmp/test type tmpfs (rw,relatime,size=4194304k,nr_inodes=102400)
````

abbradar replied:
Aha! It's actually quite interesting: it removes options when given filesystem type and device again:
```
root@abbradarbook> mount -t tmpfs tmpfs -o rw,nosuid,nodev,mode=755 testtest                                      /tmp
root@abbradarbook> mount | grep testtest                                                                          /tmp
tmpfs on /tmp/testtest type tmpfs (rw,nosuid,nodev,relatime,mode=755)
root@abbradarbook> mount -o remount,size=3046460k testtest                                                        /tmp
root@abbradarbook> mount | grep testtest                                                                          /tmp
tmpfs on /tmp/testtest type tmpfs (rw,nosuid,nodev,relatime,size=3046460k,mode=755)
root@abbradarbook> mount -t tmpfs tmpfs -o remount,size=3046460k testtest                                         /tmp
root@abbradarbook> mount | grep testtest                                                                          /tmp
tmpfs on /tmp/testtest type tmpfs (rw,relatime,size=3046460k,mode=755)
```

abbradar replied:
Also drops options when given just a device. So to preserve all the options you need to avoid mentioning both device and filesystem type.

dezgeg replied:
Ah, yes, that seems intentional according to manual:

````
              The remount functionality follows the standard way the mount command works with options from fstab.  This means that mount does not read fstab (or mtab) only when both device and dir are specified.

                  mount -o remount,rw /dev/foo /dir

              After this call all old mount options are replaced and arbitrary stuff from fstab (or mtab) is ignored, except the loop= option which is internally generated and maintained by the mount command.

                  mount -o remount,rw  /dir

              After this call, mount reads fstab and merges these options with the options from the command line (-o).  If no mountpoint is found in fstab, then a remount with unspecified source is allowed.
````

abbradar replied:
Regardless, I still think that resetting options for all special file systems here is a good idea because with systemd update or otherwise may come new mount options that we can set without rebooting.

dezgeg replied:
Hm, would we actually need to remount here ourselves if we just added the special filesystems to the normal `/etc/fstab` as well, and let `nixos-rebuild` do the remounting if needed?

dezgeg replied:
Meh, apparently not, the code explicitly refuses to create any mount units referring to them.

abbradar replied:
Yeah, I've tried to do it like this first but got multiple errors after rebuild about nonexistent `.mount` units. 

On August 28, 2016 8:53:44 PM GMT+03:00, Tuomas Tynkkynen <notifications@github.com> wrote:
>> @@ -154,9 +154,15 @@ in
>>  
>>      system.activationScripts.tmpfs =
>>        ''
>> -        ${pkgs.utillinux}/bin/mount -o
>"remount,size=${config.boot.devSize}" none /dev
>
>Meh, apparently not, the code explicitly refuses to create any mount
>units referring to them.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/NixOS/nixpkgs/pull/17822/files/58d8347ba7d5a517a03fd5209a9e8483dbdea225#r76536291

-- 
Nikolay. 

Sent from a mobile device, please excuse my brevity.

abbradar commented:
I wanted to add some option, like `special`, to `fileSystems` instead but decided against it for two reasons:

* Most options in `fileSystems` won't be supported for those mount points ;
* This option needs to be `listOf` because order matters here; we don't have anything that can mount them intelligently in the right order like systemd early.
--
Nikolay. 

Sent from a mobile device, please excuse my brevity.

abbradar commented:
I want to also rewrite hidepid module to set needed options here -- it would be much cleaner and preserve our mount options (which hidepid would now lose on remount). However, it would require making `boot.specialFileSystems` a set after all to be able to add an option, and thus #11484 (toposort).

cc @joachifm

joachifm commented:
On Fri, Aug 19, 2016, at 03:41 PM, Nikolay Amiantov wrote:
> I want to also rewrite hidepid module to set needed options here -- it
> would be much cleaner and preserve our mount options (which hidepid would
> now lose on remount). However, it would require making
> `boot.specialFileSystems` a set after all to be able to add an option,
> and thus #11484 (toposort).
> 
> cc @joachifm 

As far as I can ascertain, hidepid is not lost on remount. +1 on getting
rid of the systemd unit, though, as long as functionality is retained.

abbradar commented:
@joachifm I got confused by `mount` behaviour; see https://github.com/NixOS/nixpkgs/pull/17822#discussion_r75466140. Still think it would be nice to set it from the start without a service, yes.

abbradar committed:
nixos filesystems: unify early filesystems handling

A new internal config option `fileSystems.<name>.early` is added to indicate
that the filesystem needs to be loaded very early (i.e. in initrd). They are
transformed to a shell script in `system.build.earlyMountScript` with calls to
an undefined `specialMount` function, which is expected to be caller-specific.
This option is used by stage-1, stage-2 and activation script to set up and
remount those filesystems.  Options for them are updated according to systemd
defaults.

abbradar commented:
Updated this on top of #11484. This is much cleaner now! I've dropped separate `boot.specialFileSystems` and instead used an attribute (we can do it now with toposort). I also made hidepid to use this -- it's a lot better now.

I'll leave this for several days for discussion and merge then.

domenkozar commented:
This is a good candidate why [ova fails to build](http://hydra.nixos.org/build/39888565/nixlog/10/raw).

Diff that broke ova generation: http://hydra.nixos.org/api/scmdiff?type=git&uri=https%3A%2F%2Fgithub.com%2FNixOS%2Fnixpkgs.git&rev2=986e64280e40edb7b03e40216204e2f7fbe4e396&rev1=6b20d5b75efb3e4c95f7dabbfe0e607c1205f9be&branch=

abbradar commented:
I'll look, thanks for the heads up!

On September 4, 2016 2:33:21 PM GMT+03:00, "Domen Kožar" <notifications@github.com> wrote:
>This is a good candidate why [ova fails to
>build](http://hydra.nixos.org/build/39888565/nixlog/10/raw).
>
>Diff that broke ova generation:
>http://hydra.nixos.org/api/scmdiff?type=git&uri=https%3A%2F%2Fgithub.com%2FNixOS%2Fnixpkgs.git&rev2=986e64280e40edb7b03e40216204e2f7fbe4e396&rev1=6b20d5b75efb3e4c95f7dabbfe0e607c1205f9be&branch=
>
>-- 
>You are receiving this because you modified the open/close state.
>Reply to this email directly or view it on GitHub:
>https://github.com/NixOS/nixpkgs/pull/17822#issuecomment-244598190

-- 
Nikolay. 

Sent from a mobile device, please excuse my brevity.

domenkozar commented:
@abbradar let me know how that goes. This is the last blocker to get 16.09 channel update. cc https://github.com/NixOS/nixpkgs/issues/18209

dezgeg commented:
It's this part:
````
finalising the installation...
updating GRUB 2 menu...
installing the GRUB 2 boot loader on /dev/sda...
Installing for i386-pc platform.
Installation finished. No error reported.
setting up /etc...
mount: mount point /dev/pts does not exist
````

Presumably that's the remount part in the activation script. Probably this `runInLinuxVM` stuff should mount `/dev/pts` as well.

abbradar commented:
I think I've fixed it in 698cadd714ed19ec463e3845be2aa87718521eaf. BTW, building is very slow D:

abbradar commented:
Oh, and thanks @dezgeg for finding the cause!

domenkozar commented:
Please also pick to 16.06

On Sun, Sep 4, 2016, 16:12 Nikolay Amiantov <notifications@github.com>
wrote:

> Oh, and thanks @dezgeg <https://github.com/dezgeg> for finding the cause!
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/17822#issuecomment-244605645>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAHtg8D7iq4r6j7-TCo84EsGvY0qr8Isks5qmtG4gaJpZM4JniEY>
> .
>

abbradar commented:
If you mean 16.09 I've done this -- see 1dc5f92487e718cd9caf1c2eaa320294a1d366af

domenkozar commented:
@abbradar can I ask you to write a changelog for this PR?

abbradar commented:
Sure!

aszlig commented on nixos/modules/tasks/filesystems.nix:
```diff
@@ -97,11 +99,22 @@ let
+        internal = true;
+        description = ''
+	  Mount this filesystem very early during boot. At the moment of
+	  mounting no disks are exposed, so this option is primarily for
```
Inconsistent indentation.

abbradar replied:
This option was later removed and replaced by `boot.specialFileSystems` set.

abbradar commented:
@domenkozar Done in b2c827c]
